service/history/historyEngine.go

```diff
@@ -1031,7 +1035,25 @@ func (e *historyEngineImpl) getMutableStateOrPolling(
 		if err != nil {
 			return nil, err
 		}
-		timer := time.NewTimer(e.shard.GetConfig().LongPollExpirationInterval(domainCache.GetInfo().Name))
+
+		expirationInterval := e.shard.GetConfig().LongPollExpirationInterval(domainCache.GetInfo().Name)
+		if deadline, ok := ctx.Deadline(); ok {
+			timeout := deadline.Sub(e.shard.GetTimeSource().Now())
+			if timeout < minPollMutableStateTimeout {
+				// timeout is too short, return directly
+				// note that minPollMutableStateTimeout is larger than pollMutableStateBufferDuration,
+				// this ensure that:
+				//     1. timeout-pollMutableStateBufferDuration > 0
+				//     2. leaves a reasonable amount of time to poll mutable state
+				return response, nil
```

Comment:
In this case, I think we have 3 options:
1. return current response and nil error. when client side see the empty response, it will keep calling the API with a shorter timeout, and eventually leads to a timeout. This will impact SLA and result in more API calls.
2. return timeout error, will impact our SLA if user workflow can't complete within the timeout they initially specified.
3. return an application error, like BadRequest. My understanding is that user will see this error, and will be confused.

Without change on the client side (to stop calling server when the timeout is very short), none of them seems to be a good option. Any thoughts on it? @meiliang86 Thanks.

Reply:
2: this means the SLA is bad and should be fixed, not that we should consider adding complexity to the code to work around it.

Reply:
I agree. That's why in the comment I suggest making changes to the client side instead of here so that it won't call this API with an unreasonable small context timeout. But I guess I still need the same logic for 1 (making sure the timeout is > buffer duration so that we can return an empty response just before the context timeout).

The issue right now is that the client side logic will keep calling this API will smaller and smaller timeout which almost guarantees a timeout error and affect our SLA if the initial context timeout they specified on the client side is less than their workflow execution time.